### PR TITLE
Universal Build for Big Sur

### LINKED ISSUE #18049: Needs universal framework build on macOS Big Sur
##### System information (version)

- OpenCV => 4.4.0
- Operating System / Platform => macOS 11.0
- Compiler => CMake

##### Detailed description

I found a temporary workaround to build arm64 opencv2.framework on the [DTK](https://developer.apple.com/programs/universal/) with Apple Silicon. And cross build should works but the try failed on the intel machine. (Intel and Apple Silicon could build arm64 & x86_64. Both could run x86_64 (via Rosetta). But arm64 running is Apple Silicon only).

Prepare:
- opencv-4.4.0
- opencv_contrib-4.4.0
- native cmake from: `brew install -s cmake`

With:
- arm64 DTK 
- macOS 11.0 Beta 4
- Xcode 12 Beta 4

Build:
1. Change the [build_framework.py](https://github.com/opencv/opencv/blob/c3bb57afeaf030f10939204d48d7c2a3842f4293/platforms/osx/build_framework.py#L62) line 62: "x86_64" to "arm64"
2. `% ./opencv-4.4.0/platforms/osx/build_framework.py ./opencv-framework-4.4.0 --contrib ./opencv_contrib-4.4.0 --enable_nonfree --disable CAROTENE`
3. get opencv2.framework output

Test:
1. `% ./opencv-4.4.0/platforms/osx/run_tests.py --framework_dir=/Users/mainasuk/Developer/opencv-framework-4.4.0 --framework_name=opencv2 /Users/mainasuk/Developer/opencv-framework-4.4.0/build/build-arm64-macosx/modules/objc/test`
2. fail with error: "Unable to find a destination matching the provided destination specifier: { platform:macOS, arch:x86_64 }"
3. open OpenCVTest.xcodeproj under `/Users/mainasuk/Developer/opencv-framework-4.4.0/build/build-arm64-macosx/modules/objc/test_build`
4. In the Build Settings set project Architectures to "Standard Architectures(64-bit Intel and ARM) - $(ARCHS_STANDARD)"
5. Run unit tests get compile error "Type of expression is ambiguous without more context" on line 27

```
    func testPoint2dToMat() {
        let pointsIn = [Point2d(x:3.80004, y:73.2), Point2d(x:16.01, y:7.1111), Point2d(x:3.14, y:6), Point2d(x:-78, y:14)]
        let pointsOut = Converters.Mat_to_vector_Point2d(Converters.vector_Point2d_to_Mat(pointsIn))
        XCTAssertEqual(pointsIn, pointsOut)    <<< here
    }
```

6. The `pointsIn` with type `[Point2d]` and pointsOut with `[Point2f]`
7. Comment out the line 27
8. All unit tests pass (skip testPoint2dToMat)

I have a Swift port [repo](https://github.com/MainasuK/OpenCVBridge) for OpenCV. After create folder "Frameworks" under the root and move the opencv2.framework into it. All demo program in `OpenCVBridge.xcodeproj` run identically as the Intel Mac.

And here are 3 problems for that. 
1. Disable CAROTENE fix arm64 build fail. I have no idea about that so temporary disable it. Maybe it's could be fixed.
2. The universal cross build framework needs with x86_64 and arm64 archs. Rosetta works for Intel app but arm64 arch is required for a native build.
3. Fix needs for testPoint2dToMat unit test. :)

I hope that transform will not so hard.

### PR DESCRIPTION
**Pull request by komakai**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work

This PR resolves #18049
- add `--archs` parameter to the MacOS build
- add handling for cross-compiling cases x86_64 -> arm64 and arm64 -> x86_64

### CODE REVIEW

File: platforms/osx/build_framework.py [hunk 1]
```diff
@@ -54,13 +55,16 @@ def getInfoPlist(self, builddirs):
...
+
     if args.legacy_build:
         args.framework_name = "opencv2"
         if not "objc" in args.without:
             args.without.append("objc")
 
     b = OSXBuilder(args.opencv, args.contrib, False, False, args.without, args.disable, args.enablenonfree,
         [
-            (["x86_64"], "MacOSX")
+            (archs, "MacOSX")
```
Comment (alalek) [inline platforms/osx/build_framework.py:68]: Please use this form here:
```
("x86_64", "MacOSX"),
("arm64", "MacOSX"),
```
(as result)

Comment (komakai) [reply platforms/osx/build_framework.py:68]: OSXBuilder shares code with the iOSBuilder so we need to pass the parameters in a way that is consistent with how it is done on iOS:
```
[
    (["armv7, "armv7s", "arm64"], "iPhoneOS"),
    (["x86_64", "i386"], "iPhoneSimulator"),
]
```

Comment (alalek) [reply platforms/osx/build_framework.py:68]: Right.
I have confused by `arch = archs[0]` line in `makeCMakeCmd()`.

For each arch we should have own dedicated CMake pass.
It is impossible to "merge" them - current CMake scripts are targeted to single configuration at once.

Comment (komakai) [reply platforms/osx/build_framework.py:68]: > I have confused by arch = archs[0] line in makeCMakeCmd()

OK I understand what you mean - I can fix that

**Review by alalek — APPROVED** [review]
Well done! Thank you for contribution :+1:

**Pull request merged**